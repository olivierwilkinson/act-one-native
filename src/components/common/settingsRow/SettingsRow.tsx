import React from "react";
import { TouchableHighlight, View } from "react-native";
import { Ionicons } from "@expo/vector-icons";
import styled from "styled-components/native";

import {
  smallSizeFont,
  mediumSizeFont,
  subFont
} from "../../../styles/typography";
import { mediumLightGray, darkGray } from "../../../styles/colours";

const SettingLeftIcon = styled(Ionicons)`
  margin: 0 20px;
`;

const SettingView = styled.View`
  display: flex;
  flex-direction: row;
  align-items: center;
  padding: 10px 0px;
  background: white;
  height: 70px;
  width: 100%;
`;

const SettingContentView = styled.View`
  flex: 1;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  height: 100%;
  padding-right: 10px;

  border-bottom-width: 1px;
  border-bottom-color: ${mediumLightGray};
`;

const SettingRightView = styled.View`
  flex: 3;
  display: flex;
  flex-direction: row;
  justify-content: flex-end;
  align-items: center;
  position: relative;
  margin-left: auto;
`;

const SettingRightArrow = styled(Ionicons)`
  margin-left: 10px;
`;

const LabelText = styled.Text`
  ${mediumSizeFont}
  color: ${darkGray};
`;

const ValueText = styled.Text`
  ${subFont}
  ${smallSizeFont}
  color: gray;
`;

export type Props = {
  label: string;
  onPress?: () => void;
  value?: string;
  leftIconName?: string;
};

export default ({ label, onPress, leftIconName, value }: Props) => {
  return (
    <TouchableHighlight onPress={onPress}>
      <SettingView>
        {leftIconName && (
          <View testID="settings-row-left-icon">
            <SettingLeftIcon size={24} name={leftIconName} />
          </View>
        )}

        <SettingContentView>
          <LabelText>{label}</LabelText>

          <SettingRightView>
            {!!value && <ValueText numberOfLines={1}>{value}</ValueText>}

            {onPress && (
              <View testID="settings-row-right-arrow">
                <SettingRightArrow
                  size={20}
                  name="ios-arrow-forward"
                  color="gray"
                />
              </View>
            )}
          </SettingRightView>
        </SettingContentView>
      </SettingView>
    </TouchableHighlight>
  );
};
