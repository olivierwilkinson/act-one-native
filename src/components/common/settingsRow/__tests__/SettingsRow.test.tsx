import "react-native";
import React from "react";
import {
  render,
  RenderAPI,
  fireEvent,
  GetByAPI
} from "react-native-testing-library";

import SettingsRow, { Props } from "../SettingsRow";

describe("SettingsRow", () => {
  let defaultProps: Props;
  let getByText: GetByAPI["getByText"];
  let getByTestId: GetByAPI["getByTestId"];
  let toJSON: RenderAPI["toJSON"];
  let rerender: RenderAPI["rerender"];
  beforeEach(() => {
    defaultProps = {
      label: "Setting label"
    };

    ({ getByText, getByTestId, toJSON, rerender } = render(
      <SettingsRow {...defaultProps} />
    ));
  });

  it("renders correctly", () => {
    expect(toJSON()).toMatchSnapshot();
  });

  describe("when passed onPress", () => {
    beforeEach(() => {
      defaultProps = {
        ...defaultProps,
        onPress: jest.fn()
      };

      rerender(<SettingsRow {...defaultProps} />);
    });

    it("renders right arrow", () => {
      expect(getByTestId("settings-row-right-arrow")).not.toBeNull();
    });

    it("calls onPress when pressed", () => {
      fireEvent.press(getByText(defaultProps.label));

      expect(defaultProps.onPress).toHaveBeenCalled();
    });
  });

  describe("when passed leftIconName", () => {
    beforeEach(() => {
      defaultProps = {
        ...defaultProps,
        leftIconName: "airplane"
      };

      rerender(<SettingsRow {...defaultProps} />);
    });

    it("renders left icon", () => {
      expect(getByTestId("settings-row-left-icon")).not.toBeNull();
    });
  });
});
