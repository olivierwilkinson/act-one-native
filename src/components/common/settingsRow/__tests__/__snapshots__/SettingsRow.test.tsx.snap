// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsRow renders correctly 1`] = `
<View
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "white",
          "display": "flex",
          "flexDirection": "row",
          "height": 70,
          "paddingBottom": 10,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 10,
          "width": "100%",
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "alignItems": "center",
            "borderBottomColor": "rgb(240, 240, 240)",
            "borderBottomWidth": 1,
            "display": "flex",
            "flexBasis": 0,
            "flexDirection": "row",
            "flexGrow": 1,
            "flexShrink": 1,
            "height": "100%",
            "justifyContent": "space-between",
            "paddingRight": 10,
          },
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#202020",
              "fontSize": 16,
            },
          ]
        }
      >
        Setting label
      </Text>
      <View
        style={
          [
            {
              "alignItems": "center",
              "display": "flex",
              "flexBasis": 0,
              "flexDirection": "row",
              "flexGrow": 3,
              "flexShrink": 1,
              "justifyContent": "flex-end",
              "marginLeft": "auto",
              "position": "relative",
            },
          ]
        }
      />
    </View>
  </View>
</View>
`;
